cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -101,6 +111,60 @@ private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail, Id.
     // return the first one from the artifact details list.
     return artifactDetails.get(0).getDescriptor();
   }
+  /**
+   * A RemovalListener for closing classloader and plugin instantiators.
+   */
+  private static final class InstantiatorsRemovalListener implements
+    RemovalListener<ArtifactDescriptor, Instantiators> {
+
+    @Override
+    public void onRemoval(RemovalNotification<ArtifactDescriptor, Instantiators> notification) {
+      Closeables.closeQuietly(notification.getValue());
+    }
+  }
+
+  private class Instantiators implements Closeable {
+    private final CloseableClassLoader parentClassLoader;
+    private final Map<ArtifactDescriptor, PluginInstantiator> pluginInstantiatorMap;
```

Comment:
this isn't thread safe, may get concurrent modification errors if parallel calls are made.